src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1080,6 +1087,9 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                         }
                     }
                 }
+            } else if (var_annotation == "S") {
+                type = ASRUtils::TYPE(ASR::make_SymbolicExpression_t(al, loc));
+                return type;
```

Comment:
Should this be one `if-else` level before, that is parallel to `if-else` for `pointer`, `bool`, `str`, etc.? Currently, it seems it is in the `if-else` for user-defined types like `Struct`, `Enum`.  Since `S` here seems to be an internal type, maybe it could make more sense to have this check parallel to the existing internal type checks.

PS: I guess this can be done in a separate PR.

Reply:
Ohh, this was something Thirumalai and I discussed over . Check this thread out https://github.com/lcompilers/lpython/pull/1846#discussion_r1228155048
We've kept it below `Struct` as `S` , the internal type can also be defined by the user and in that case we would be using it as a user defined type !

Reply:
It might probably be fine for now. I think a better approach could be where we resolve the `type_annotation` at the start of the function `get_type_from_var_annotation` and handle appropriately. For example, instead of moving `else if (var_annotation == "S") ` inside, how about if we resolve the annotation first and bring `if (s) {` outside (probably make it the first `if`) parallel to other internal type checks. That way it also works for other predefined/internal types.

For example, diff with respect to main branch:
```bash
$ git diff
diff --git a/src/lpython/semantics/python_ast_to_asr.cpp b/src/lpython/semantics/python_ast_to_asr.cpp
index 0c1783d1a..f529afa9a 100644
--- a/src/lpython/semantics/python_ast_to_asr.cpp
+++ b/src/lpython/semantics/python_ast_to_asr.cpp
@@ -1005,7 +1005,31 @@ public:
         AST::expr_t** m_args=nullptr, size_t n_args=0,
         bool raise_error=true) {
         ASR::ttype_t* type = nullptr;
-        if (var_annotation == "i8") {
+        ASR::symbol_t *s = current_scope->resolve_symbol(var_annotation);
+        if (s) {
+            if (ASR::is_a<ASR::Variable_t>(*s)) {
+                ASR::Variable_t *var_sym = ASR::down_cast<ASR::Variable_t>(s);
+                if (var_sym->m_type->type == ASR::ttypeType::TypeParameter) {
+                    ASR::TypeParameter_t *type_param = ASR::down_cast<ASR::TypeParameter_t>(var_sym->m_type);
+                    type = ASRUtils::TYPE(ASR::make_TypeParameter_t(al, loc, type_param->m_param));
+                    return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
+                }
+            } else {
+                ASR::symbol_t *der_sym = ASRUtils::symbol_get_past_external(s);
+                if( der_sym ) {
+                    if ( ASR::is_a<ASR::StructType_t>(*der_sym) ) {
+                        type = ASRUtils::TYPE(ASR::make_Struct_t(al, loc, s));
+                        return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
+                    } else if( ASR::is_a<ASR::EnumType_t>(*der_sym) ) {
+                        type = ASRUtils::TYPE(ASR::make_Enum_t(al, loc, s));
+                        return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
+                    } else if( ASR::is_a<ASR::UnionType_t>(*der_sym) ) {
+                        type = ASRUtils::TYPE(ASR::make_Union_t(al, loc, s));
+                        return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
+                    }
+                }
+            }
+        } else if (var_annotation == "i8") {
             type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc, 1));
             type = ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
         } else if (var_annotation == "i16") {
@@ -1055,32 +1079,10 @@ public:
             bool is_allocatable = false;
             type = ast_expr_to_asr_type(underlying_type->base.loc, *underlying_type, is_allocatable);
             type = ASRUtils::TYPE(ASR::make_Pointer_t(al, loc, type));
+        } else if (var_annotation == "S") {
+                type = ASRUtils::TYPE(ASR::make_SymbolicExpression_t(al, loc));
+                return type;
         } else {
-            ASR::symbol_t *s = current_scope->resolve_symbol(var_annotation);
-            if (s) {
-                if (ASR::is_a<ASR::Variable_t>(*s)) {
-                    ASR::Variable_t *var_sym = ASR::down_cast<ASR::Variable_t>(s);
-                    if (var_sym->m_type->type == ASR::ttypeType::TypeParameter) {
-                        ASR::TypeParameter_t *type_param = ASR::down_cast<ASR::TypeParameter_t>(var_sym-
>m_type);
-                        type = ASRUtils::TYPE(ASR::make_TypeParameter_t(al, loc, type_param->m_param));
-                        return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
-                    }
-                } else {
-                    ASR::symbol_t *der_sym = ASRUtils::symbol_get_past_external(s);
-                    if( der_sym ) {
-                        if ( ASR::is_a<ASR::StructType_t>(*der_sym) ) {
-                            type = ASRUtils::TYPE(ASR::make_Struct_t(al, loc, s));
-                            return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
-                        } else if( ASR::is_a<ASR::EnumType_t>(*der_sym) ) {
-                            type = ASRUtils::TYPE(ASR::make_Enum_t(al, loc, s));
-                            return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
-                        } else if( ASR::is_a<ASR::UnionType_t>(*der_sym) ) {
-                            type = ASRUtils::TYPE(ASR::make_Union_t(al, loc, s));
-                            return ASRUtils::make_Array_t_util(al, loc, type, dims.p, dims.size());
-                        }
-                    }
-                }
-            }
             if( raise_error ) {
                 throw SemanticError("Unsupported type annotation: " + var_annotation, loc);
             }
(END)
```